### DISCUSSION TITLE
Help on sampling from depth buffer WebGL vs. WebGPU

### DISCUSSION THREAD

**Discussion by AshleyScirra** [gpuweb/gpuweb#2277] [Q&A] [answered]
I'm trying to write a fog effect based on sampling a depth texture. In WebGL I have this shader code to sample depth and linearize it:

```glsl
mediump float depthSample = texture2D(samplerDepth, depthUV).r;
mediump float zLinear = zNear * zFar / (zFar + depthSample * (zNear - zFar));
```

In WGSL I have:

```wgsl
var depthSample : f32 = textureSample(textureDepth, samplerDepth, depthUV);
var zLinear : f32 = zNear * zFar / (zFar + depthSample * (zNear - zFar));
```
where `textureDepth` is `texture_depth_2d`.

As far as I can tell everything on the API side is configured the same in both cases. But these calculations seem to produce a different `zLinear` value, resulting in a different fog effect between WebGL and WebGPU (testing in Chrome Canary). From what I can find WebGL ought to be normalizing depth samples to 0-1, and I assumed WebGPU would do the same, but it's pretty hopeless trying to trawl the spec to find a reference on this. Am I missing something?

**Comment by AshleyScirra** [comment, marked as answer]
OK so with a bit of trial and error I found this WGSL code matches the GLSL:

```wgsl
var depthSample : f32 = textureSample(textureDepth, samplerDepth, depthUV);
depthSample = (depthSample + 1.0) / 2.0;
var zLinear : f32 = zNear * zFar / (zFar + depthSample * (zNear - zFar));
```

with the extra line being `depthSample = (depthSample + 1.0) / 2.0;`. In other words it looks like WebGPU is returning a depth sample in the range [-1, 1], whereas I was expecting one in [0, 1]. Is that a bug?

**Reply by Kangz** [reply]
I don't think that the spec says anything about this yet, but it definitely seems like a bug since the WebGPU clip space goes from 0 to 1  for Z (I think?) can you file it on crbug.com/dawn ?

**Reply by Kangz** [reply]
Also nice catch figuring the issue out!

**Reply by kainino0x** [reply]
Agree this sounds like an implementation bug to me. I would also expect it to return [0,1].

**Reply by AshleyScirra** [reply]
Filed: https://bugs.chromium.org/p/dawn/issues/detail?id=1187
